### ISSUE TITLE
ActiveRecord leaves sleeping threads after connection.close

### ISSUE THREAD

**Issue by cpetley** [rails/rails#36333]
I am seeing an increase in used threads after connections are made and then closed. 

```ruby
start_thread_count = Thread.list.count

User.establish_connection
User.connection.close

end_thread_count = Thread.list.count
added_thread_count = end_thread_count - start_thread_count

puts "Added thread count: #{added_thread_count}"
// -> 1
```

In Rails 5.1.7 this code has no increase in threads (which I believe is the correct/expected behaviour?), but Rails 5.2 onwards has this issue. This results in tens of thousands of sleeping threads being used for connection_pool until eventually no more threads can be spawned. Restarting the app releases the threads. 

**Rails version**: 5.2.2

**Ruby version**:2.5.1

**Comment by cpetley** [comment]
This issue looks to be related #36330

**Comment by rafaelfranca** [comment]
Can you try in the master branch?

**Comment by rafaelfranca** [comment]
cc @jhawthorn since you worked on the reapers. Do you think you commit fixed this problem?

**Comment by jhawthorn** [comment]
👍 Yes. That commit should mean we won't leak threads (but if we're leaking connection pool objects that's also bad)

**Comment by jhawthorn** [comment]
Thread leak is avoided by #36296, and the more fundamental issue is fixed by #36345

**Comment by cpetley** [comment]
@jhawthorn thanks very much for your work on this - much appreciated. Will this fix be included in the Rails 5.2.4 release, or only R6? We can't upgrade to Rails 6 yet and we're trying to decide whether or not to roll back or wait for R5.2.4. Thanks again
